perl/modules/Zulip/lib/BarnOwl/Module/Zulip.pm

```diff
@@ -0,0 +1,522 @@
+use strict;
+use warnings;
+
+package BarnOwl::Module::Zulip;
+
+our $VERSION=0.1;
+our $queue_id;
+our $last_event_id;
+our %cfg;
+our $max_retries = 1000;
+our $retry_timer;
+our $tls_ctx;
+our %msg_id_map;
+our $presence_timer;
+
+use AnyEvent;
+use AnyEvent::HTTP;
+use AnyEvent::TLS;
+use JSON;
+use MIME::Base64;
+use URI;
+use URI::Encode;
+use BarnOwl::Hooks;
+use BarnOwl::Message::Zulip;
+use HTTP::Request::Common;
+use Getopt::Long qw(GetOptionsFromArray);
+
+sub fail {
+    my $msg = shift;
+    undef %cfg;
+    undef $queue_id;
+    undef $last_event_id;
+
+    BarnOwl::admin_message('Zulip Error', $msg);
+    die("Zulip Error: $msg\n");
+}
+
+
+sub initialize {
+    my $conffile = BarnOwl::get_config_dir() . "/zulip";
+
+    if (open(my $fh, "<", "$conffile")) {
+        read_config($fh);
+        close($fh);
+    }
+    if ($cfg{'api_url'} =~ /^https/) {
+        if (exists $cfg{'ssl_key_file'}) {
+            $tls_ctx = new AnyEvent::TLS(verify => $cfg{'ssl_verify'}, 
+                                         sslv3 => 0, verify_peername => "http",
+                                         ca_file => $cfg{'ssl_ca_file'},
+                                         cert_file => $cfg{'ssl_cert_file'},
+                                         key_file => $cfg{'ssl_key_file'});
+        } else {
+            $tls_ctx = new AnyEvent::TLS(verify => $cfg{'ssl_verify'}, 
+                                         sslv3 => 0, verify_peername => "http",
+                                         ca_file => $cfg{'ssl_ca_file'});
+        }           
+    } else {
+        # we still want it for a unique id
+        $tls_ctx = int(rand(1024));
```

Comment:
How unique can it be if you expect a collision within ~32 calls?

Reply:
I honestly don't remember why I picked 1024 here. Do you think replacing it with "high" (the AnyEvent::TLS "validate certs" canned context) is sufficient?